cli/flox-rust-sdk/src/models/manifest/composite/mod.rs

```diff
@@ -0,0 +1,246 @@
+/// contents of the low priority vector.
+fn shallow_merge_optional_vecs<T: Clone>(
+    vec_lp: Option<&Vec<T>>,
+    vec_hp: Option<&Vec<T>>,
+) -> Option<Vec<T>> {
+    vec_hp.cloned().or(vec_lp.cloned())
+}
+
+/// Takes the high priority `T` if it's present, otherwise the low priority `T`.
+fn shallow_merge_options<T: Clone>(opt_lp: Option<&T>, opt_hp: Option<&T>) -> Option<T> {
+    opt_hp.cloned().or(opt_lp.cloned())
+}
+
+fn deep_merge_optional_containerize_config(
+    cfg_lp: Option<&ManifestContainerizeConfig>,
+    cfg_hp: Option<&ManifestContainerizeConfig>,
+) -> Option<ManifestContainerizeConfig> {
+    match (cfg_lp, cfg_hp) {
+        (None, None) => None,
+        (Some(cfg), None) => Some(cfg.clone()),
+        (None, Some(cfg)) => Some(cfg.clone()),
+        (Some(cfg_lp), Some(cfg_hp)) => {
+            let cfg = ManifestContainerizeConfig {
+                user: shallow_merge_options(cfg_lp.user.as_ref(), cfg_hp.user.as_ref()),
+                exposed_ports: optional_set_union(
+                    cfg_lp.exposed_ports.as_ref(),
+                    cfg_hp.exposed_ports.as_ref(),
+                ),
+                cmd: shallow_merge_options(cfg_lp.cmd.as_ref(), cfg_hp.cmd.as_ref()),
+                volumes: optional_set_union(cfg_lp.volumes.as_ref(), cfg_hp.volumes.as_ref()),
+                working_dir: shallow_merge_options(
+                    cfg_lp.working_dir.as_ref(),
+                    cfg_hp.working_dir.as_ref(),
+                ),
+                labels: optional_map_union(cfg_lp.labels.as_ref(), cfg_hp.labels.as_ref()),
+                stop_signal: shallow_merge_options(
+                    cfg_lp.stop_signal.as_ref(),
+                    cfg_hp.stop_signal.as_ref(),
+                ),
+            };
+            Some(cfg)
+        },
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::shallow::ShallowMerger;
+    use super::*;
+    use crate::models::manifest::typed::Inner;
+
+    #[test]
+    fn composite_manifest_runs_merger() {
+        let empty = Manifest::default();
+        let manifest1 = {
+            let mut manifest = Manifest::default();
+            manifest
+                .vars
+                .inner_mut()
+                .insert("var1".to_string(), "manifest1".to_string());
+            manifest
+        };
+        let manifest2 = {
+            let mut manifest = Manifest::default();
+            manifest
+                .vars
+                .inner_mut()
+                .insert("var2".to_string(), "manifest2".to_string());
+            manifest
+        };
+        let composite = CompositeManifest {
+            composer: empty,
+            deps: vec![manifest1, manifest2],
+        };
+        let merged = composite.merge_all(ShallowMerger).unwrap();
+        assert_eq!(merged.vars.inner()["var1"], "manifest1");
+        assert_eq!(merged.vars.inner()["var2"], "manifest2");
```

Comment:
We've lost some coverage here that the `composer` is applied last and that it's preserved when there are no deps. (as an aside, I found it really hard to prove that a manifest was applied more than once, but decided that it's probably not a big deal).

Reply:
Fixed